File: 0050-powx-n/0050-powx-n.cpp
class Solution {
public:
    double f(double x, double n) {
        if(n == 0) return 1;
        
        double t = f(x, n/2);
        
        if((int)n%2 == 0) return t*t;
        
        return t*t*x;
    }
    
    double myPow(double x, int n) {
        if(n == 0) return 1;
        
        bool flag = n < 0;
        n = abs(n);
        
        auto ans = f(x, n);
        
        if(flag) return 1/ans;
        
        return ans;
    }
};

File: 0050-powx-n/0050-powx-n.go
func f(x float64, n float64) float64 {
    if n == 0 {
        return 1
    }
    
    ans := f(x, n/2)
    if int(n)%2 == 1 {
        return ans*ans*x
    }
    
    return ans*ans
}

func myPow(x float64, n int) float64 {
    flag := (n < 0)
    t := float64(n)
    t = math.Abs(t)
    
    res := f(x, t)
    if flag {
        return 1/res
    }
    
    return res
}